docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+the next evolution of ABCI, within the Cosmos SDK. ABCI++ aims to provide
+application developers with more flexibility and control over application and
+consensus semantics, e.g. in-application mempools, in-process oracles, and
+order-book style matching engines.
+
+## Context
+
+Tendermint will release ABCI++ in phases. Notably, at the time of this writing,
+Tendermint is releasing v0.37.0 which will include `PrepareProposal` and `ProcessProposal`.
+
+The `PrepareProposal` ABCI method is concerned with a block proposer requesting
+the application to evaluate a series of transactions to be included in the next
+block, defined as a slice of `TxRecord` objects. The application can either
+accept, reject, or completely ignore some or all of these transactions. This is
+an important consideration to make as the application can essentially define and
+control its own mempool allowing it to define sophisticated transaction priority
+and filtering mechanisms, by completely ignoring the `TxRecords` Tendermint
+sends it, favoring its own transactions. This essentially means that the Tendermint
+mempool acts more like a gossip data structure.
+
+The second ABCI method, `ProcessProposal`, is used to process the block proposer's
+proposal as defined by `PrepareProposal`. This ABCI method requests that the
+application evaluate the entire proposed block for validity.
+
+It is important to note that in this phase of ABCI++ integration, the application
+is NOT responsible for locking semantics -- Tendermint will still be responsible
+for that. In the future, however, the application will be responsible for locking,
+which allows for parallel execution possibilities.
+
+## Decision
+
+We will integrate the first phase of ABCI++, which will be introduced in Tendermint
+v0.37.0, in the next major release of the Cosmos SDK. We will integrate the two
+aforementioned ABCI++ methods on the `BaseApp` type. We describe the implementations
+of the two methods individually below.
+
+Prior to describing the implementation of the two new methods, it is important to
+note that the existing ABCI methods, `CheckTx`, `DeliverTx`, etc, still exist and
+serve the same functions as they do now.
+
+### `PrepareProposal`
+
+Prior to evaluating the decision for how to implement `PrepareProposal`, it is
+important to note that `CheckTx` will still be executed and will be responsible
+for evaluating transaction validity as it does now, with one very important
+_additive_ distinction.
+
+When executing transactions in `CheckTx`, the application will now add valid
+transactions, i.e. passing the AnteHandler, to it's own mempool data structure.
+In order to provide a flexible approach to meet the varying needs of application
+developers, we will define both a mempool interface and a data structure utilizing
+Golang generics, allowing developers to focus only on transaction
```

Comment:
Someone unrelated, but it's not clear to me why this needs generics. From the definition below, this could be an interface that accepts the `MempoolTx` interface in its method signatures.

Reply:
I suppose it technically doesn't :P